### PR TITLE
Upgrade `k8s.io/*` to `v0.27.5`, `sigs.k8s.io/controller-runtime` to `v0.15.2`

### PR DESCRIPTION
**Pull request by ary1992**
**How to categorize this PR?**

/area open-source dev-productivity
/kind enhancement

**What this PR does / why we need it**:
Upgrade dependencies

- k8s.io/* to v0.27.5
- sigs.k8s.io/controller-runtime to v0.15.2

**Which issue(s) this PR fixes**:
Part of #7991 
Part of https://github.com/gardener/gardener/issues/7473

**Special notes for your reviewer**:
- Using `old Restmapper` until the fix is merged upstream https://github.com/kubernetes-sigs/controller-runtime/pull/2425
- Using the `old code-generator` implementation as in the latest implementation, internal packages as not generated. Refer https://github.com/kubernetes/kubernetes/pull/115196

**Release note**:

```other developer
The following dependencies are updated:
- `k8s.io/*` : `v0.26.4` -> `v0.27.5`
- `sigs.k8s.io/controller-runtime`: `v0.14.6` -> `v0.15.2`
```
```breaking developer
`uncachedObjects` under pkg/client/kubernetes/options.go is now removed from Config struct which is used to set options for new ClientSets. Now the uncached objects can be directly set under `clientOptions.Cache.DisableFor` field.
```

### CODE REVIEW

File: extensions/pkg/controller/worker/reconciler_test.go [hunk 1]
```diff
@@ -110,9 +110,7 @@ var _ = Describe("Worker Reconcile", func() {
 	)
 
 	DescribeTable("Reconcile function", func(t *test) {
-		mgr.EXPECT().GetClient().Return(t.fields.client).AnyTimes()
-		mgr.EXPECT().GetAPIReader().Return(t.fields.client).AnyTimes()
-		reconciler := worker.NewReconciler(mgr, t.fields.actuator(ctrl))
+		reconciler := worker.NewReconciler(t.fields.actuator(ctrl))
```
Comment (shafeeqes) [inline extensions/pkg/controller/worker/reconciler_test.go]: This change is unrelated to this commit?

File: pkg/operator/controller/garden/care/add.go [hunk 2]
```diff
@@ -84,8 +84,8 @@ func (r *Reconciler) AddToManager(ctx context.Context, mgr manager.Manager) erro
 				r.Config.Controllers.GardenCare.SyncPeriod.Duration,
 			),
 		}).
-		Watches(
-			source.NewKindWithCache(&operatorv1alpha1.Garden{}, mgr.GetCache()),
+		WatchesRawSource(
+			source.Kind(mgr.GetCache(), &operatorv1alpha1.Garden{}),
```
Comment (shafeeqes) [inline pkg/operator/controller/garden/care/add.go]: You could use 
```
		Watches(
			&operatorv1alpha1.Garden{},
```
directly since we watch the manager cache itself.
See,
https://github.com/gardener/gardener/blob/ee820b4f1e4e52c3f5243ed3f1c84dd0e9ae02cf/vendor/sigs.k8s.io/controller-runtime/pkg/builder/controller.go#L132-L140

File: cmd/gardener-resource-manager/app/app.go [hunk 3]
```diff
@@ -210,8 +209,6 @@ func run(ctx context.Context, log logr.Logger, cfg *config.ResourceManagerConfig
 				return apiutil.NewDynamicRESTMapper(
 					config,
 					httpClient,
-					apiutil.WithLazyDiscovery,
-					apiutil.WithLimiter(rate.NewLimiter(rate.Every(1*time.Minute), 1)), // rediscover at maximum every minute
```
Comment (shafeeqes) [inline cmd/gardener-resource-manager/app/app.go:212]: Was this replaced with some new functionality, or?

File: extensions/pkg/webhook/certificates/reloader.go [hunk 4]
```diff
@@ -58,7 +59,7 @@ type reloader struct {
 // manager in order to periodically reload the secret from the cluster.
 func (r *reloader) AddToManager(ctx context.Context, mgr manager.Manager) error {
 	r.reader = mgr.GetClient()
-	r.certDir = mgr.GetWebhookServer().CertDir
+	r.certDir = mgr.GetWebhookServer().(*webhook.DefaultServer).Options.CertDir
```
Comment (shafeeqes) [inline extensions/pkg/webhook/certificates/reloader.go]: Should we do error handling here, just to be safe?

File: pkg/component/extensions/extension/extension_test.go [hunk 5]
```diff
@@ -165,6 +164,8 @@ var _ = Describe("Extension", func() {
 			},
 		}
 
+		fakeSeedClient = fakeclient.NewClientBuilder().WithScheme(kubernetes.SeedScheme).WithStatusSubresource(beforeExtension).Build()
```
Comment (shafeeqes) [inline pkg/component/extensions/extension/extension_test.go]:
```suggestion
		fakeSeedClient = fakeclient.NewClientBuilder().WithScheme(kubernetes.SeedScheme).WithStatusSubresource(&extensionsv1alpha1.Extension{}).Build()
```
Similarly, for others.

File: extensions/pkg/controller/worker/reconciler_test.go [hunk 6]
```diff
@@ -110,7 +110,9 @@ var _ = Describe("Worker Reconcile", func() {
 	)
 
 	DescribeTable("Reconcile function", func(t *test) {
-		reconciler := worker.NewReconciler(t.fields.actuator(ctrl))
+		mgr.EXPECT().GetClient().Return(t.fields.client).AnyTimes()
+		mgr.EXPECT().GetAPIReader().Return(t.fields.client).AnyTimes()
```
Comment (shafeeqes) [inline extensions/pkg/controller/worker/reconciler_test.go]:
```
apiReader = mockclient.NewMockReader(ctrl)
mgr.EXPECT().GetAPIReader().Return(apiReader)
```

**Review by shafeeqes — COMMENTED** [review]
Well-structured PR, some initial comments.

**Comment by shafeeqes** [comment]
You can also vendor `v0.27.4` directly and update envtest version here, https://github.com/gardener/gardener/blob/74020278eabec42dd107577462844b43b1b857ff/hack/prepare-envtest.sh#L21

**Comment by rfranzke** [comment]
/unassign
due to vacation

File: extensions/pkg/webhook/certificates/reconciler.go [hunk 7]
```diff
@@ -86,7 +86,12 @@ type reconciler struct {
 // AddToManager generates webhook CA and server cert if it doesn't exist on the cluster yet. Then it adds reconciler to
 // the given manager in order to periodically regenerate the webhook secrets.
 func (r *reconciler) AddToManager(ctx context.Context, mgr manager.Manager) error {
-	r.serverPort = mgr.GetWebhookServer().(*webhook.DefaultServer).Options.Port
+	defaultServer, ok := mgr.GetWebhookServer().(*webhook.DefaultServer)
+	if !ok {
+		return fmt.Errorf("expected *webhook.DefaultServer, got %T", defaultServer)
```
Comment (shafeeqes) [inline extensions/pkg/webhook/certificates/reconciler.go]: For printing the type, you should use the original object, see 
https://github.com/gardener/gardener/blob/5abed1bfedd6f2188e5dcec9af161d1b5c286236/pkg/resourcemanager/webhook/seccompprofile/handler.go#L36-L39

**Comment by rfranzke** [comment]
Since k/k@v1.28 is released, should we directly update to that and ctrl-runtime v0.16?

**Comment by ary1992** [comment]
> Since k/k@v1.28 is released, should we directly update to that and ctrl-runtime v0.16?

Sure

**Comment by shafeeqes** [comment]
> Since k/k@v1.28 is released, should we directly update to that and ctrl-runtime v0.16?

Isn't this PR too big already, there might be more breaking changes with the new dependencies. I would vote for a new PR as it would also help during revendoring in extensions.

File: docs/api-reference/core.md [hunk 8]
```diff
@@ -7549,8 +7549,8 @@ See <a href="https://github.com/kubernetes/ingress-nginx/blob/master/docs/user-g
 <td>
 <code>externalTrafficPolicy</code></br>
 <em>
-<a href="https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#serviceexternaltrafficpolicytype-v1-core">
-Kubernetes core/v1.ServiceExternalTrafficPolicyType
+<a href="https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#serviceexternaltrafficpolicy-v1-core">
```
Comment (shafeeqes) [inline docs/api-reference/core.md]:
```suggestion
<a href="https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#serviceexternaltrafficpolicy-v1-core">
```
Use 1.27 for generating these

File: go.mod [hunk 9]
```diff
@@ -42,36 +42,39 @@ require (
...
+	k8s.io/metrics v0.27.4
+	k8s.io/pod-security-admission v0.27.4
+	k8s.io/utils v0.0.0-20230209194617-a36077c30491
 	sigs.k8s.io/controller-runtime v0.14.6
 	sigs.k8s.io/controller-runtime/tools/setup-envtest v0.0.0-20221212190805-d4f1e822ca11 // v0.14.1
 	sigs.k8s.io/controller-tools v0.11.3
 	sigs.k8s.io/yaml v1.3.0
 )
 
+require sigs.k8s.io/structured-merge-diff/v4 v4.2.3 // indirect
```
Comment (rfranzke) [inline go.mod]: Can you add this to the same `require` block below?

File: extensions/pkg/webhook/certificates/reconciler.go [hunk 10]
```diff
@@ -85,7 +86,7 @@ type reconciler struct {
 // AddToManager generates webhook CA and server cert if it doesn't exist on the cluster yet. Then it adds reconciler to
 // the given manager in order to periodically regenerate the webhook secrets.
 func (r *reconciler) AddToManager(ctx context.Context, mgr manager.Manager) error {
-	r.serverPort = mgr.GetWebhookServer().Port
+	r.serverPort = mgr.GetWebhookServer().(*webhook.DefaultServer).Options.Port
```
Comment (rfranzke) [inline extensions/pkg/webhook/certificates/reconciler.go]: In the release notes they write:

> The following webhook server options have been deprecated, use manager.WebhookServer instead.
manager.Options.Port
manager.Options.Host
manager.Options.CertDir
manager.Options.TLSOpts

Can you adapt? Same for other occurrences

File: cmd/gardenlet/app/app.go [hunk 11]
```diff
@@ -268,14 +268,21 @@ func (g *garden) Start(ctx context.Context) error {
 				},
 			}
 
+			ns := gardenerutils.ComputeGardenNamespace(g.config.SeedConfig.SeedTemplate.Name)
```
Comment (rfranzke) [inline cmd/gardenlet/app/app.go]: call this `seedNamespace`

File: cmd/gardenlet/app/app.go [hunk 12]
```diff
@@ -268,14 +268,21 @@ func (g *garden) Start(ctx context.Context) error {
...
 			return kubernetes.AggregatorCacheFunc(
 				kubernetes.NewRuntimeCache,
 				map[client.Object]cache.NewCacheFunc{
 					// Gardenlet should watch secrets only in the seed namespace of the seed it is responsible for. We
 					// don't use any selector mechanism here since we want to still fall back to reading secrets with
 					// the API reader (i.e., not from cache) in case the respective secret is not found in the cache.
-					&corev1.Secret{}:         cache.MultiNamespacedCacheBuilder([]string{gardenerutils.ComputeGardenNamespace(g.config.SeedConfig.SeedTemplate.Name)}),
-					&corev1.ServiceAccount{}: cache.MultiNamespacedCacheBuilder([]string{gardenerutils.ComputeGardenNamespace(g.config.SeedConfig.SeedTemplate.Name)}),
+					&corev1.Secret{}: func(c *rest.Config, o cache.Options) (cache.Cache, error) {
+						o.Namespaces = []string{ns, ns}
```
Comment (rfranzke) [inline cmd/gardenlet/app/app.go]: Add a comment why we add the namespace twice.

File: cmd/gardenlet/app/app.go [hunk 13]
```diff
@@ -268,14 +268,21 @@ func (g *garden) Start(ctx context.Context) error {
...
+						return cache.New(c, o)
+					},
+					&corev1.ServiceAccount{}: func(c *rest.Config, o cache.Options) (cache.Cache, error) {
+						o.Namespaces = []string{ns, ns}
```
Comment (rfranzke) [inline cmd/gardenlet/app/app.go]: same here

File: docs/development/kubernetes-clients.md [hunk 14]
```diff
@@ -212,7 +212,7 @@ _Important characteristics of cached controller-runtime clients:_
 In order to allow more granular control over which object kinds should be cached and which calls should bypass the cache, controller-runtime offers a few mechanisms to further tweak the client/cache behavior:
 
 - When creating a `DelegatingClient`, certain object kinds can be configured to always be read directly from the API instead of from the cache. Note that this does not prevent starting a new Informer when retrieving them directly from the cache.
-- Watches can be restricted to a given (set of) namespace(s) by using `cache.MultiNamespacedCacheBuilder` or setting `cache.Options.Namespace`.
+- Watches can be restricted to a given (set of) namespace(s) by setting `cache.Options.Namespace`.
```
Comment (rfranzke) [inline docs/development/kubernetes-clients.md]: Namespace*s*, right?

File: cmd/gardener-resource-manager/app/app.go [hunk 15]
```diff
@@ -206,9 +207,10 @@ func run(ctx context.Context, log logr.Logger, cfg *config.ResourceManagerConfig
 			// use dynamic rest mapper for target cluster, which will automatically rediscover resources on NoMatchErrors
 			// but is rate-limited to not issue to many discovery calls (rate-limit shared across all reconciliations)
 			opts.MapperProvider = func(config *rest.Config, httpClient *http.Client) (meta.RESTMapper, error) {
-				return apiutil.NewDynamicRESTMapper(
+				return thirdpartyapiutil.NewDynamicRESTMapper(
```
Comment (rfranzke) [inline cmd/gardener-resource-manager/app/app.go:217]: Can you link the related PR here (https://github.com/kubernetes-sigs/controller-runtime/pull/2425) with a TODO statement that this should be reverted to `apiutil` once this PR is available? Same for other occurrences

File: go.mod [hunk 16]
```diff
@@ -61,9 +61,9 @@ require (
 	k8s.io/metrics v0.27.4
 	k8s.io/pod-security-admission v0.27.4
 	k8s.io/utils v0.0.0-20230209194617-a36077c30491
-	sigs.k8s.io/controller-runtime v0.15.0
+	sigs.k8s.io/controller-runtime v0.15.1
```
Comment (rfranzke) [inline go.mod]: there is 0.15.2

File: pkg/component/plutono/plutono_test.go [hunk 17]
```diff
@@ -784,6 +785,10 @@ func testDashboardConfigMap(ctx context.Context, c client.Client, namespaceName
 		availableDashboards = sets.Set[string]{}
 	)
 
+	cmList := &corev1.ConfigMapList{}
+	ExpectWithOffset(1, c.List(ctx, cmList)).To(Succeed())
+	fmt.Println(cmList.Items[0].Name)
```
Comment (rfranzke) [inline pkg/component/plutono/plutono_test.go]: remove this?

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: cmd/gardener-admission-controller/app/app.go [hunk 18]
```diff
@@ -117,6 +118,12 @@ func run(ctx context.Context, log logr.Logger, cfg *config.AdmissionControllerCo
 		MetricsBindAddress:     net.JoinHostPort(cfg.Server.Metrics.BindAddress, strconv.Itoa(cfg.Server.Metrics.Port)),
 
 		LeaderElection: false,
+
+		WebhookServer: controllerwebhook.NewServer(controllerwebhook.Options{
```
Comment (rfranzke) [inline cmd/gardener-admission-controller/app/app.go:118]: Remove L113-115 now?

File: cmd/gardenlet/app/app.go [hunk 19]
```diff
@@ -284,19 +287,27 @@ func (g *garden) Start(ctx context.Context) error {
...
 				map[client.Object]cache.NewCacheFunc{
 					// Gardenlet should watch secrets only in the seed namespace of the seed it is responsible for. We
 					// don't use any selector mechanism here since we want to still fall back to reading secrets with
 					// the API reader (i.e., not from cache) in case the respective secret is not found in the cache.
 					&corev1.Secret{}: func(c *rest.Config, o cache.Options) (cache.Cache, error) {
-						o.Namespaces = []string{ns, ns}
+						// cache.New only creates a multiNamespacedCache if cache.Options.Namespaces has more than one namespace
+						// (https://github.com/kubernetes-sigs/controller-runtime/blob/116a1b831fffe7ccc3c8145306c3e1a3b1b14ffa/pkg/cache/cache.go#L202-L204).
+						// MultiNamespacedCache can only be possible when there are more than one namespace
+						// (https://github.com/kubernetes-sigs/controller-runtime/blob/116a1b831fffe7ccc3c8145306c3e1a3b1b14ffa/pkg/cache/multi_namespace_cache.go#L53-L55)
```
Comment (rfranzke) [inline cmd/gardenlet/app/app.go]: Can you also add an explanation why we need a multi namespaced cache here?

**Comment by shafeeqes** [comment]
As discussed in https://github.com/gardener/gardener/pull/8217#discussion_r1257819054 for worker controller,
Can you also do the same for https://github.com/gardener/gardener/blob/bc94ce9bd342f3c18f206018edeca7278fcb4cfc/pkg/provider-local/controller/controlplane/add.go#L51-L54, Let's return an error for `NewActuator`. This was missed in #8217.

File: cmd/gardenlet/app/app.go (hunk 19 above)
Comment (ary1992) [reply cmd/gardenlet/app/app.go]: Above line https://github.com/gardener/gardener/blob/5c71cb98dcf958a31b583fec75d22e16b0b04322/cmd/gardenlet/app/app.go#L294-L296 already explains it right?

File: pkg/resourcemanager/controller/secret/add.go [hunk 20]
```diff
@@ -52,8 +52,8 @@ func (r *Reconciler) AddToManager(ctx context.Context, mgr manager.Manager, sour
 			// with too many unnecessary requests for all secrets in cluster/namespace.
 			resourcemanagerpredicate.HasFinalizer(r.ClassFilter.FinalizerName()),
 		)).
-		Watches(
-			&source.Kind{Type: &resourcesv1alpha1.ManagedResource{}},
+		WatchesRawSource(
+			source.Kind(mgr.GetCache(), &resourcesv1alpha1.ManagedResource{}),
```
Comment (acumino) [inline pkg/resourcemanager/controller/secret/add.go]:
```suggestion
		Watches(
			&resourcesv1alpha1.ManagedResource{},
```

File: go.mod [hunk 21]
```diff
@@ -22,48 +22,49 @@ require (
...
+	k8s.io/apiextensions-apiserver v0.27.4
+	k8s.io/apimachinery v0.27.4
+	k8s.io/apiserver v0.27.4
 	k8s.io/autoscaler/vertical-pod-autoscaler v0.14.0
-	k8s.io/client-go v0.26.3
-	k8s.io/cluster-bootstrap v0.26.3
-	k8s.io/code-generator v0.26.3
-	k8s.io/component-base v0.26.3
-	k8s.io/component-helpers v0.26.3
+	k8s.io/client-go v0.27.4
```
Comment (rfranzke) [inline go.mod]: Use v0.27.5

Comment (shafeeqes) [reply go.mod]: There is https://github.com/kubernetes-sigs/controller-runtime/releases/tag/v0.15.2 as well

File: cmd/gardenlet/app/app.go [hunk 22]
```diff
@@ -306,23 +306,21 @@ func (g *garden) Start(ctx context.Context) error {
 		// read a secret from another namespace. There might be secrets in namespace other than the seed-specific
 		// namespace (e.g., backup secret in the SeedSpec). Hence, let's use a fallback client which falls back to an
 		// uncached reader in case it fails to read objects from the cache.
-		opts.NewClient = func(cache cache.Cache, config *rest.Config, options client.Options, uncachedObjects ...client.Object) (client.Client, error) {
-			uncachedClient, err := client.New(config, options)
+		opts.NewClient = func(config *rest.Config, options client.Options) (client.Client, error) {
+			uncachedOptions := options
+			uncachedOptions.Cache = nil
+			uncachedClient, err := client.New(config, uncachedOptions)
```
Comment (acumino) [inline cmd/gardenlet/app/app.go:337]: Here for creating an uncached client `option.Cache` is specifically set to nil, should the same be done [here](https://github.com/gardener/gardener/blob/54b0182e8f7231ca7b5de88bf179fb6dc4192851/pkg/client/kubernetes/client.go#L289-L295)?

File: pkg/client/kubernetes/client.go [hunk 23]
```diff
@@ -302,18 +309,20 @@ func newClientSet(conf *Config) (Interface, error) {
...
-					Reader:     runtimeCache,
-					DisableFor: conf.uncachedObjects,
-				},
-			})
+			cacheOptions := conf.clientOptions.Cache
+			if cacheOptions == nil {
+				cacheOptions = &client.CacheOptions{}
+			}
+			cacheOptions.Reader = runtimeCache
+			conf.clientOptions.Cache = cacheOptions
```
Comment (rfranzke) [inline pkg/client/kubernetes/client.go]: You could extract this into a function and call it here and in the place above

**Review by rfranzke — CHANGES_REQUESTED** [review]
Did you check whether we can keep calling the `generate-{internal-}groups.sh` scripts from the `k8s.io/code-generator` repository and don't need to copy them?

File: pkg/client/kubernetes/client.go [hunk 24]
```diff
@@ -378,6 +364,23 @@ func defaultContentTypeProtobuf(c *rest.Config) *rest.Config {
...
+
+	cacheOptions.Reader = reader
+	conf.clientOptions.Cache = cacheOptions
+
+	client, err := client.New(conf.restConfig, conf.clientOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	return client, nil
```
Comment (rfranzke) [inline pkg/client/kubernetes/client.go]:
```suggestion
	return client.New(conf.restConfig, conf.clientOptions)
```

**Review by rfranzke — COMMENTED** [review]
/lgtm
/cc @shafeeqes @acumino

**Comment by ary1992** [comment]
> Did you check whether we can keep calling the `generate-{internal-}groups.sh` scripts from the `k8s.io/code-generator` repository and don't need to copy them?

We cannot simply pass the `internalversion` to `CLIENTSET_NAME_VERSIONED` variable for the `generate-{internal-}groups.sh`. There were more changes done to the script which needs more insight. I prefer to do it as a part of separate PR.

**Review by shafeeqes — COMMENTED** [review]
Thank you!
/lgtm

**Review by acumino — COMMENTED** [review]
/approve
